tldextract/utils.py

```diff
@@ -22,12 +23,16 @@ def return_cache(*args, **kwargs):
 
             cache_path = path + '/' + key + '.json'
             cache_path = cache_path.replace('//', '/')
+            lock_path = cache_path + '.lock'
 
-            if not os.path.isfile(cache_path):
-                result = func(*args, **kwargs)
-                make_dir(cache_path)
-                with open(cache_path, 'w') as cache_file:
-                    json.dump(result, cache_file)
+            make_dir(cache_path)
+
+            # without locking, Concurrency could lead to multiple writers, or readers of a partial write
+            with FileLock(lock_path, timeout=20):
+                if not os.path.isfile(cache_path):
+                    result = func(*args, **kwargs)
+                    with open(cache_path, 'w') as cache_file:
+                        json.dump(result, cache_file)
 
             with open(cache_path) as cache_file:
                 return json.load(cache_file)
```

Comment:
shouldn't the read also be protected by the lock?

Reply:
good catch will fix